### ISSUE TITLE
The alerts from renaming a folder in Whodata Windows doesn't prompt properly in the redesign of FIM

### ISSUE THREAD

**Issue by Skeptor** [wazuh/wazuh#4321]
|Related issue|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|---|
|https://github.com/wazuh/wazuh/issues/3319| 3.11 | Syscheck | Agent | Sources | Windows |

Since Wazuh version 3.9.1 (https://github.com/wazuh/wazuh/pull/2888), Syscheck will show 2 alerts for each file (Remove and add) after renaming a monitored subfolder. Those alerts have audit information and are prompt in real-time.

```
** Alert 1575381928.436000: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2019 Dec 03 14:05:28 (win-3.10) any->syscheck
Rule: 553 (level 7) -> 'File deleted.'
File 'c:\users\pablo\desktop\test-whodata\cambio_de_nombre\new text document.txt' was deleted.

Attributes:
 - (Audit) User name: pablo
 - (Audit) Process id: 2144
 - (Audit) Process name: C:\Windows\explorer.exe

** Alert 1575381928.436436: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2019 Dec 03 14:05:28 (win-3.10) any->syscheck
Rule: 554 (level 5) -> 'File added to the system.'
File 'c:\users\pablo\desktop\test-whodata\name_change\new text document.txt' was added.

Attributes:
 - Size: 0
 - Permissions: SYSTEM (allowed): delete|read_control|write_dac|write_owner|synchronize|read_data|write_data|append_data|read_ea|write_ea|execute|read_attributes|write_attributes, Administrators (allowed): delete|read_control|write_dac|write_owner|synchronize|read_data|write_data|append_data|read_ea|write_ea|execute|read_attributes|write_attributes, pablo (allowed): delete|read_control|write_dac|write_owner|synchronize|read_data|write_data|append_data|read_ea|write_ea|execute|read_attributes|write_attributes
 - Date: Tue Dec  3 13:28:11 2019
 - User: pablo (S-1-5-21-1087345056-2530757087-3950806551-1000)
 - MD5: d41d8cd98f00b204e9800998ecf8427e
 - SHA1: da39a3ee5e6b4b0d3255bfef95601890afd80709
 - SHA256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
 - File attributes: ARCHIVE

```

After the redesign of FIM, those changes are detected during the scheduled scan, and the alerts don't show any audit information.
The function `w_update_sacl` is unused in the redesign of FIM and it was the function that enables this behaviour.

```
** Alert 1575381240.419980: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2019 Dec 03 13:54:00 (windows) any->syscheck
Rule: 554 (level 5) -> 'File added to the system.'
File 'c:\users\pablo\desktop\poc_replace\namechanged\rootcheck.template' added
Mode: scheduled

Attributes:
 - Size: 51
 - Permissions: SYSTEM (allowed): delete|read_control|write_dac|write_owner|synchronize|read_data|write_data|append_data|read_ea|write_ea|execute|read_attributes|write_attributes, Administradores (allowed): delete|read_control|write_dac|write_owner|synchronize|read_data|write_data|append_data|read_ea|write_ea|execute|read_attributes|write_attributes, pablo (allowed): delete|read_control|write_dac|write_owner|synchronize|read_data|write_data|append_data|read_ea|write_ea|execute|read_attributes|write_attributes
 - Date: Mon Oct 14 07:34:37 2019
 - User: pablo (S-1-5-21-4181897701-2688429609-2145492800-1001)
 - MD5: 24c1f37a73d24835fa9cdf9a8b6c7ab5
 - SHA1: 445c7a1b80b9778e0a7209ae0c4df599ab527d29
 - SHA256: 7b749eafcf988cc2dd41b765579493fe5123e36c0e035d38f3593a93cb979159
 - File attributes: ARCHIVE

** Alert 1575381240.421142: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2019 Dec 03 13:54:00 (windows) any->syscheck
Rule: 553 (level 7) -> 'File deleted.'
File 'c:\users\pablo\desktop\poc_replace\name_changed\rootcheck.template' deleted
Mode: scheduled

Attributes:
 - Size: 320
 - Permissions: SYSTEM (allowed): delete|read_control|write_dac|write_owner|synchronize|read_data|write_data|append_data|read_ea|write_ea|execute|read_attributes|write_attributes, Administradores (allowed): delete|read_control|write_dac|write_owner|synchronize|read_data|write_data|append_data|read_ea|write_ea|execute|read_attributes|write_attributes, pablo (allowed): delete|read_control|write_dac|write_owner|synchronize|read_data|write_data|append_data|read_ea|write_ea|execute|read_attributes|write_attributes
 - Date: Mon Oct 14 07:34:56 2019
 - User: pablo (S-1-5-21-4181897701-2688429609-2145492800-1001)
 - MD5: 929cbc238a06e0575e7401455fe565ca
 - SHA1: 0413085e26973d1d048f27b8f8c9640caa4c86cb
 - SHA256: 4cd11fcc5434b8c7653b48c0051f7cfa36016f5101453d50e7ec1ab5fe253388
 - File attributes: ARCHIVE
```

**Comment by crolopez** [comment]
We have verified that the event related to the renaming of directories monitored in whodata for Linux systems doesn't work either. The reason for this is the fact that the Audit engine does not detect the creation events that should be triggered after a directory rename.

The events that Audit generates when performing a directory rename are:

```
type=PROCTITLE msg=audit(1576066094.588:85): proctitle=6D7600666F6C646572332F00666F6C64657234
type=SYSCALL msg=audit(1576066094.588:85): arch=c000003e syscall=82 success=yes exit=0 a0=7fff09e338a9 a1=7fff09e338b2 a2=562092574550 a3=642 items=5 ppid=14577 pid=14617 auid=1000 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=pts3 ses=11 comm="mv" exe="/bin/mv" key="wazuh_fim"
type=CWD msg=audit(1576066094.588:85): cwd="/root/test"
type=PATH msg=audit(1576066094.588:85): item=0 name="/root/test" inode=272275 dev=08:03 mode=040755 ouid=0 ogid=0 rdev=00:00 nametype=PARENT
type=PATH msg=audit(1576066094.588:85): item=1 name="/root/test" inode=272275 dev=08:03 mode=040755 ouid=0 ogid=0 rdev=00:00 nametype=PARENT
type=PATH msg=audit(1576066094.588:85): item=2 name=(null) inode=272275 dev=08:03 mode=040755 ouid=0 ogid=0 rdev=00:00 nametype=PARENT
type=PATH msg=audit(1576066094.588:85): item=3 name=(null) inode=268490 dev=08:03 mode=040755 ouid=0 ogid=0 rdev=00:00 nametype=DELETE
type=PATH msg=audit(1576066094.588:85): item=4 name="folder4" inode=268490 dev=08:03 mode=040755 ouid=0 ogid=0 rdev=00:00 nametype=CREATE
```

However, unlike whodata for Windows, this has never worked in previous versions.

Since `3.9.1`, we are processing directory renames when the whodata thread receives a directory remove. Then, a directory scan is performed over the folder which contained the removed directory (does not have to match the configured or "parent" directory). Finally, the hash table is iterated to send a remove event for each file from the deleted directory.

https://github.com/wazuh/wazuh/blob/7aa4d46661f8b0f97b0969634482626f30866731/src/syscheckd/win_whodata.c#L1747-L1783

If we change the data structure and want to implement this functionality in whodata for Windows, it will be necessary to implement a mechanism to navigate through data structure and correct it.

**Comment by vikman90** [comment]
Fixed by #4557.
